staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go

```diff
@@ -471,31 +477,184 @@ func TestUpdateNodesInExternalLoadBalancer(t *testing.T) {
+				{Service: newService("s4", "123", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1}},
+			},
+			worker:                3,
+			nodeListerErr:         nil,
+			expectedRetryServices: []*v1.Service{},
+		},
+		{
+			desc:  "2 nodes",
+			nodes: []*v1.Node{node1, node2},
+			expectedUpdateCalls: []fakecloud.UpdateBalancerCall{
+				{Service: newService("s0", "777", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2}},
+				{Service: newService("s1", "888", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2}},
+				{Service: newService("s3", "999", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2}},
+				{Service: newService("s4", "123", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2}},
+			},
+			worker:                1,
+			nodeListerErr:         nil,
+			expectedRetryServices: []*v1.Service{},
+		},
+		{
+			desc:  "4 nodes",
+			nodes: []*v1.Node{node1, node2, node3, node4},
+			expectedUpdateCalls: []fakecloud.UpdateBalancerCall{
+				{Service: newService("s0", "777", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2, node3, node4}},
+				{Service: newService("s1", "888", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2, node3, node4}},
+				{Service: newService("s3", "999", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2, node3, node4}},
+				{Service: newService("s4", "123", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2, node3, node4}},
+			},
+			worker:                3,
+			nodeListerErr:         nil,
+			expectedRetryServices: []*v1.Service{},
+		},
+		{
+			desc:                  "error occur during sync",
+			nodes:                 []*v1.Node{node1, node2, node3, node4},
+			expectedUpdateCalls:   []fakecloud.UpdateBalancerCall{},
+			worker:                3,
+			nodeListerErr:         fmt.Errorf("random error"),
+			expectedRetryServices: services,
+		},
+		{
+			desc:                  "error occur during sync with 1 worker",
+			nodes:                 []*v1.Node{node1, node2, node3, node4},
+			expectedUpdateCalls:   []fakecloud.UpdateBalancerCall{},
+			worker:                1,
+			nodeListerErr:         fmt.Errorf("random error"),
+			expectedRetryServices: services,
+		},
+	} {
+		controller.nodeLister = newFakeNodeLister(tc.nodeListerErr, tc.nodes...)
+		servicesToRetry := controller.updateLoadBalancerHosts(services, tc.worker)
+		compareServiceList(t, tc.desc, tc.expectedRetryServices, servicesToRetry)
+		compareUpdateCalls(t, tc.desc, tc.expectedUpdateCalls, cloud.UpdateCalls)
+		cloud.UpdateCalls = []fakecloud.UpdateBalancerCall{}
+	}
+}
+
+// compareServiceList compares if both left and right inputs contains the same service list despite the order.
+func compareServiceList(t *testing.T, desc string, left, right []*v1.Service) {
+	if len(left) != len(right) {
+		t.Errorf("for case %q, expect len(left) == len(right), but got %v != %v", desc, len(left), len(right))
+	}
```

Comment:
Actually - on a second though, is this really needed?
Can't we just:
- sort both lists
- and compare using DeepEqual
?

Reply:
Yes. That would achieve the same effect. It is just that needs to add code to implement the interfaces needed for sorting. This struct would just be used for testing. So I am not sure if that is worth it.